src/redis-benchmark.c

```diff
@@ -506,44 +513,39 @@ static void readHandler(aeEventLoop *el, int fd, void *privdata, int mask) {
                     exit(1);
                 }
                 redisReply *r = reply;
-                int is_err = (r->type == REDIS_REPLY_ERROR);
-
-                if (is_err && config.showerrors) {
-                    /* TODO: static lasterr_time not thread-safe */
-                    static time_t lasterr_time = 0;
-                    time_t now = time(NULL);
-                    if (lasterr_time != now) {
-                        lasterr_time = now;
+                if (r->type == REDIS_REPLY_ERROR) {
+                    /* Try to update slots configuration if reply error is
+                    * MOVED/ASK/CLUSTERDOWN and the key(s) used by the command
+                    * contain(s) the slot hash tag. */
+                    if (c->cluster_node && c->staglen) {
+                        int fetch_slots = 0, do_wait = 0;
+                        if (!strncmp(r->str,"MOVED",5) || !strncmp(r->str,"ASK",3))
+                            fetch_slots = 1;
+                        else if (!strncmp(r->str,"CLUSTERDOWN",11)) {
+                            /* Usually the cluster is able to recover itself after
+                            * a CLUSTERDOWN error, so try to sleep one second
+                            * before requesting the new configuration. */
+                            fetch_slots = 1;
+                            do_wait = 1;
+                            printf("Error from server %s:%d: %s\n",
+                                c->cluster_node->ip,
+                                c->cluster_node->port,
+                                r->str);
+                        }
+                        if (do_wait) sleep(1);
+                        if (fetch_slots && !fetchClusterSlotsConfiguration(c))
+                            exit(1);
```

Comment:
it's a little bit hard to review this diff. IIUC, it didn't really change (other than the comment below.
what changed is that it went inside an additional separate `if` (on `r->type == REDIS_REPLY_ERROR` instead of is_err`).
and the first part at the stop about `showerrors` moved to the bottom (as an `else` without an `if`).

so logically, the only changes in that huge diff block is that:
1. we don't have the possibility to print the error twice.
2. normally MOVED, and ASK will not print any error
3. we `exit` on any unexpected error (in addition to the failure to fetch cluster slots configuration)

let me know if there's anything else i'm missing that changed in that function.

p.s. maybe there was a possibility to arrange the code in a way that it'll create a smaller diff, but never mind now.

Reply:
@oranagra thank you for the fast review. with regards to:
> 1. we don't have the possibility to print the error twice. 
correct.
> 2. normally MOVED, and ASK will not print any error
Missed it. Fixed it on the latest commit. We now print the MOVED and ASK same a before
> 3. we exit on any unexpected error (in addition to the failure to fetch cluster slots configuration)
correct.
----
> let me know if there's anything else i'm missing that changed in that function.
All covered above.
> p.s. maybe there was a possibility to arrange the code in a way that it'll create a smaller diff, but never mind now.
I also wanted to reduce the changes to the minimal but the identation change makes the line number to be large.

Reply:
do we wanna print an error on MOVED and ASK?
i suppose it'll be printed only once (per thread?) since we re-fetch the cluster configuration, right?

note another difference is that the double errors in the past were throttled to be printed once per second, now we don't have double error print cases, but also no throttling.

* the `exit` works as a way of "throttling" the unexpected errors.
* MOVED and ASK are maybe solved by fetchClusterSlotsConfiguration (so they won't flood the console)
  @filipecosta90 can you test to confirm that?
* and CLUSTERDOWN is in addition to that throttled by the sleep

Reply:
@oranagra you're totally right on  flooding the console.
 So my question is: 
 - should we print the MOVED/ASK? If yes then we need the double errors in the past were throttled to be printed once per second. If not then we're ok with the previous print position. WDYT?
```
$ ../../src/redis-benchmark -r 100000 -n 1000000000 -t set -p 30002 --cluster
Cluster has 3 master nodes:

Master 0: ff62a57dec6aae7bf257f32fc11f709c6a6529d5 127.0.0.1:30001
Master 1: b4e30ae77b444361619b8ec3d182c1e5056c2939 127.0.0.1:30002
Master 2: 7c2512ef5932a70149343780a555ff95bdde9a6c 127.0.0.1:30003

Error from server 127.0.0.1:30002: MOVED 5461 127.0.0.1:30001. 0.194)3))
Updating slots configuration.
WARNING: Cluster slots configuration changed, fetching new one...
Error from server 127.0.0.1:30002: MOVED 5461 127.0.0.1:30001.
Updating slots configuration.
Error from server 127.0.0.1:30002: MOVED 5461 127.0.0.1:30001.
Updating slots configuration.
Error from server 127.0.0.1:30002: MOVED 5461 127.0.0.1:30001.
Updating slots configuration.
Error from server 127.0.0.1:30002: MOVED 5461 127.0.0.1:30001.
Updating slots configuration.
Error from server 127.0.0.1:30002: MOVED 5461 127.0.0.1:30001.
Updating slots configuration.
Error from server 127.0.0.1:30002: MOVED 5461 127.0.0.1:30001.
Updating slots configuration.
Error from server 127.0.0.1:30002: MOVED 5461 127.0.0.1:30001.
Updating slots configuration.
Error from server 127.0.0.1:30002: MOVED 5461 127.0.0.1:30001.
Updating slots configuration.
Error from server 127.0.0.1:30002: MOVED 5461 127.0.0.1:30001.
Updating slots configuration.
Error from server 127.0.0.1:30002: MOVED 5461 127.0.0.1:30001.
Updating slots configuration.
Error from server 127.0.0.1:30002: MOVED 5461 127.0.0.1:30001.
Updating slots configuration.
Error from server 127.0.0.1:30002: MOVED 5461 127.0.0.1:30001.
Updating slots configuration.
Error from server 127.0.0.1:30002: MOVED 5461 127.0.0.1:30001.
Updating slots configuration.
Error from server 127.0.0.1:30002: MOVED 5461 127.0.0.1:30001.
Updating slots configuration.
Error from server 127.0.0.1:30002: MOVED 5461 127.0.0.1:30001.
Updating slots configuration.
Error from server 127.0.0.1:30002: MOVED 5461 127.0.0.1:30001.
Updating slots configuration.
```

Reply:
i think that MOVED and ASK are part of the protocol, and are handled correctly by the benchmark.
they can be completely silenced IMHO.
in fact, before this PR, they were also completely silence by default, right? (when the `showerrors` was false)

Reply:
@oranagra fixed per PR review: printing only CLUSTERDOWN error message.